Implemented edit functionality allowing users to update records

l00188531 opened the pull request:
![Screenshot 2025-03-18 145711](https://github.com/user-attachments/assets/11f91f57-393d-4363-8300-9a05e50234c3)

l00188531 committed:
Implemented edit functionality allowing users to update records

l00188531 commented:
The next step is to Ensure that only authorized users can edit applications

L00137900 reviewed (changes requested):
Hi Khurram, Just a few things on this one from my end:

1. Is there any reason why you choose to use PyMySQL over SQLAlchemy? I was thinking it would be better to keep consistent with the use of SQLAlchemy which we have already implemented. 
2. Duplicate environment variable loading, these are already declared in the database configuration (line 47) so it is redundant we don't need to load it twice.
3. Could add error handling around if the connection fails to establish.  
4. Could add error handling for a record when that ID does not exist for instance if someone has deleted the ID with that entry in the database and you try to edit it on the front end.

l00188531 committed:
refactored edit function to use SQLAlchemy

l00188531 commented:
there was no specific reason for using PyMySQL. I had tested these functions separately, that's why I initially missed maintaining consistency. I have refactored the code to use SQLAlchemy, removed the duplicate variable, and improved error handling for connection failures. However, I did not implement error handling for the ID, as it is a primary key and must exist in the database.

L00137900 commented:
Ok nice work on those addressing those 3 points though, on point 4 which I raised, there's no guarantee that the record hasn't been deleted after it's loaded in the front end so we need error handling implemented here just best practice

L00137900 commented:
![image](https://github.com/user-attachments/assets/0858c2f4-1349-4e90-bbdd-e976b06e38b7)
Also it is failing in SonarQube here

l00188531 committed:
implemented additional error handling in case record it deleted  before updating

l00188531 commented:
To handle the case where the record might be deleted between the time it's loaded and the time the user submits the form, I have implemented additional error handling before updating the record, SonarQube is warning about the security risk of allowing both GET and POST methods on the /edit/<int:id>. To fix the warning Next I will be implementing Authentication and Authorization Checks where the logged-in user/admin will have access to the edit path/function. hopefully, then that warning will go away

L00137900 committed:
Add CSRF protection to resolve SonarQube security warning

L00137900 commented:
- Added CSRF token protection to prevent cross-site request forgery attacks against application
- Moved cancel button from using oneclick attribute to data-url attribute to reduce the risk of cross site scripting 

I thought this might help with resolving issue in SonarQube but still seeing this issue, anyways security has improved slightly

L00137900 commented:
- Added CSRF token protection to prevent cross-site request forgery attacks against application
- Moved cancel button from using oneclick attribute to data-url attribute to reduce the risk of cross site scripting

Will resolve issue with SonarQube now in a moment

L00137900 committed:
Split GET and POST methods for edit endpoint

L00137900 commented:
Separated GET/POST operations for the edit endpoint and can see it is no longer failing in SonarQube. All good to merge.

L00137900 reviewed (approved):
All good to approve now with changes made. Functionality will be added to ensure that only authorized users can edit applications at a later stage but for now looks good!
